libraries/AP_ADSB/AP_ADSB.cpp

```diff
@@ -496,36 +505,82 @@ void AP_ADSB::handle_adsb_vehicle(const adsb_vehicle_t &vehicle)
     }
 }
 
-/*
- * Update the vehicle list. If the vehicle is already in the
- * list then it will update it, otherwise it will be added.
- */
-void AP_ADSB::handle_vehicle(const mavlink_message_t &packet)
-{
-    adsb_vehicle_t vehicle {};
-    const uint32_t now = AP_HAL::millis();
-    mavlink_msg_adsb_vehicle_decode(&packet, &vehicle.info);
-    vehicle.last_update_ms = now - (vehicle.info.tslc * 1000);
-    handle_adsb_vehicle(vehicle);
-}
-
 /*
  * Copy a vehicle's data into the list
  */
 void AP_ADSB::set_vehicle(const uint16_t index, const adsb_vehicle_t &vehicle)
 {
-    if (index >= in_state.list_size) {
+    if (index >= in_state.list_size_allocated) {
         // out of range
         return;
     }
-    in_state.vehicle_list[index] = vehicle;
+
+    uint16_t flags = vehicle.info.flags;
+
+    if (flags == 0) {
+        // if no flags set, assume most for real global hardware. This is to allow systems
+        // that don't populate any flag because maybe they don't have a way to verify it.
+        flags = ADSB_FLAGS_VALID_COORDS |
+                ADSB_FLAGS_VALID_ALTITUDE |
+                ADSB_FLAGS_VALID_HEADING |
+                ADSB_FLAGS_VALID_VELOCITY |
+                ADSB_FLAGS_VALID_CALLSIGN |
+                ADSB_FLAGS_VALID_SQUAWK |
+                ADSB_FLAGS_VERTICAL_VELOCITY_VALID |
+                ADSB_FLAGS_BARO_VALID;
+                // all flags set except ADSB_FLAGS_SOURCE_UAT and ADSB_FLAGS_SIMULATED
+    }
+
+    in_state.vehicle_list[index].last_update_ms = vehicle.last_update_ms;
+
+    // use |= so we allow partial sets and those don't override previous valid settings
+    mavlink_adsb_vehicle_t &adsb_vehicle = in_state.vehicle_list[index].info;
+    adsb_vehicle.flags |= flags;
```

Comment:
Doesn't this mean that if a flag gets turned off by the external vehicle (e.g. they've lost their position) we won't notice that?

Reply:
yes, and then the uAvionix will send us zeros for lat/lng (for better or worse). But I agree we should catch that case on our end just in case

Reply:
Are you going to patch that, then?

If we are passing these flags onto a GCS, haven't we lost the information that the original vehicle is no longer (e.g.) giving a position estimate?

Reply:
Yeah. Gimme a couple hours.

Reply:
Yes. Although I'm tempted to undo this just to get this damn thing in.